### ISSUE TITLE
Organized deprecations flow

### ISSUE THREAD

**Issue by medikoo** [serverless/serverless#7422]
As suggested in https://github.com/serverless/serverless/pull/6195#issuecomment-590249791  (cc @BryanCrotaz) it'll be nice to incorporate some safe way to introduce breaking changes to framework.

Ideally:
- Any feature to be removed (prior being removed with a new MAJOR), should be announced as breaking with a new MINOR release. (~~It might be good to require a minimum 2 months period between announcing feature as deprecated and releasing a new major that makes it effective~~).
- Deprecation notices should displayed for any command, which loads plugins and resolves user configuration. There should be no case where deprecation messages are shown for `sls deploy` but are not for e.g. `sls package `or `sls info`
- Deprecation notices for given feature should be displayed _only_ if user relies on that feature.
There should be no case where we pollute log with noisy messages which do not apply to given service.
- New MAJOR should only remove deprecated features. Therefore users which ensured to address all deprecation messages should feel safe to pursue with an upgrade.
- It should be possible to silence all or specific deprecation logs (although single warning should be outputted, notifying it's the case, and that it might unsafe to upgrade to new major).

_As I see there was once an attempt to introduce that (cc @pmuens @eahefnawy)   (https://github.com/serverless/serverless/issues/3211) but it was quickly reverted (https://github.com/serverless/serverless/pull/3418), The only outlined reason fo revert was: `it harms DX`.
Still then it was designed to output logs unconditionally, even when user was not in a scope of a service, or was in a scope of service which didn't rely on given feature. In given case indeed it could have been considered as harmful for DX_ 

### Implementation spec proposal

- Introduce `logDeprecation(code, message, { serviceConfig })` in `lib/utils/logDeprecation.js`  which should be used to log deprecations. It should be plain function free from attachment to Framework classes (as it might be used before Framework is initialized) with following logic implied:
  - Given deprecation (identified by `code`) should be logged only once. So if it's invoked repeatedly with same depracation only first invocation should produce log
  - Internally `SLS_DEPRECATION_DISABLE` env var should be resolved (possible values: `*` or `{deprecationCode1},{deprecationCode2},...{deprecationCodeN}`, and listed depracation codes should not be logged.
   - Same optional `serviceConfig` should be handled from which we resolve `serviceConfig.disabledDeprecations` (could be a single value string on array)
  - I propose following look:
![Screenshot 2020-05-14 at 15 42 17](https://user-images.githubusercontent.com/122434/81941804-95275680-95f9-11ea-9f58-2e455807291f.png)

- Each deprecation should be documented (possibly in `docs/deprecations.md`) with a clear resolution instructions. Each deprecation log should link document page and section which provides insight on given deprecation

**Comment by BryanCrotaz** [comment]
There's no need for "_a minimum 2 months period_" - as long as a user has no deprecation warnings in the last `n.minor`, it's safe to upgrade to `n+1.0`

This methodology has worked extremely successfully for the EmberJS framework.

Ember has a deprecation warning manager which makes cleaning up deprecations easy without blowing up the build log.
Some useful notes on how and why Ember does this:
https://emberjs.com/releases/
https://guides.emberjs.com/release/configuring-ember/handling-deprecations/

**Comment by AhmedFat7y** [comment]
@medikoo So we want to implement a logging function that will just log a deprecation message only when called. It will be up to the others to log the deprication message (unlike the initial approach on this #3211)?

`code` here will be text referring to specific functionality/plugin or what it will be?

Do we need to add more structure? For example: https://github.com/serverless/serverless/blob/ca2a73f91a86ae41b4cf48384177c0fd74ff4f1f/lib/classes/Error.js#L30-L35

Here you have `title` and `message`, and you print it in a defined structure, should we do the same with the `logDeprecation`? For example sth like `logDeprecation(code, message, docURL)`? Just to make the output a bit standardized ?

**Comment by medikoo** [comment]
>  It will be up to the others to log the deprication message (unlike the initial approach on this #3211)?

No it'll be just for internal use, just like in #3211. Still notable difference is that we will log deprecations only if given _deprecated_ functionality is used, and not in all cases (#3211 was reverted as it proposed logging depreciations unconditionally and that's quite bad for DX)

> So we want to implement a logging function that will just log a deprecation message only when called. 

See implementation proposal in main description (I've just updated it). I've tried to put there all necessary information. Let me know if anything remains unclear, and great thanks for interest, I think it's quite important functionality (!)

**Comment by AhmedFat7y** [comment]
I'm sorry, It's my first time deprecating anything so I'm not sure I understand the flow properly. I will be asking a lot 😄 

> No it'll be just for internal use

Let me rephrase what I meant:  If you decide to deprecate something, You will add the log where you made the change, so it will only be printed if the functionality is used. Hence, it's different from the approach in #3211 and the log messages be at random places.

For example, If we deprecate `maxAgeInSeconds` in function config, I will put the `logDeprecation` call inside `compile/function/index.js#compileFunction()` around the lines I changed

> Given deprecation (identified by code) should be logged only once. So if it's invoked repeatedly with same depracation only first invocation should produce log

This means we will store codes printed so far on the filesystem inside package directory to keep track of them through multiple calls?

**Comment by medikoo** [comment]
> You will add the log where you made the change, so it will only be printed if the functionality is used. Hence, it's different from the approach in #3211 and the log messages be at random places.

Yes, exactly

> This means we will store codes printed so far on the filesystem inside package directory to keep track of them through multiple calls?

No, it applies only to given process, so we need to just store in memory information what logs were issued. e.g. every other invocation of `serverless` should show again same logs. We just need to ensure that with one invocation we do not produce multiple logs for same thing.

**Comment by AhmedFat7y** [comment]
I want to give it a try

**Comment by medikoo** [comment]
@AhmedFat7y Thank you!

**Comment by AhmedFat7y** [comment]
@medikoo I have an implementation question for the object where I store which codes were printed so far and how to properly test it

Should I just import a function `module.exports = logDeprecation`? Then, I depend on closure variables to keep track of which codes were printed. In testing, I need to re-require the module to reset the variables in closure.

Or should I export an object and keep track of what was logged inside this object. Then this internal variable will be exposed. Testing will be much easier with stubbing, but doesn't seem that clean.


I submitted a sample implementation, let me know what you think

**Comment by medikoo** [comment]
> Should I just import a function module.exports = logDeprecation? Then, I depend on closure variables to keep track of which codes were printed

Yes, I think it should be as simple as possible,

> In testing, I need to re-require the module to reset the variables in closure.

When using `runServerless` util, you can enforce that for each `serverless` instance modules are re-initialized by passing `modulesCacheStub: {}`. So I believe that should solve that problem

**Comment by AhmedFat7y** [comment]
@medikoo should we add this as note somewhere so new contributors don't get confused?
I actually wanted to add some points for new contributors, the things that I had to look up or know from you. Pretty small things though

**Comment by medikoo** [comment]
In case of this util, I think tests you did are fine, no need to rely on `runServerless` as util is pretty top level and totally isolated from `Serverless` class

Concerning tests documentation, we already have document here: https://github.com/serverless/serverless/blob/master/tests/README.md (and it's linked in: https://github.com/serverless/serverless/blob/master/CONTRIBUTING.md)

`runServerless` is documented here: https://github.com/serverless/test/blob/master/docs/run-serverless.md but indeed `modulesCacheStub` documentation is pretty vague.

Anyway feel free to propose any improvements to above. I totally agree that there's a big room for improvements in docs.

**Comment by AhmedFat7y** [comment]
I will propose some bullet points
like for example to use `stdout.write` instead of `console.log`, or this should be left for PR Reviews? As people will try to match what exists already

**Comment by medikoo** [comment]
> I will propose some bullet points

It's a great idea! I've started to group such points here: https://github.com/serverless/serverless/blob/master/CONTRIBUTING.md#other-guideliness

**Comment by BryanCrotaz** [comment]
Check out the very successful deprecations implementation in Ember.js
https://github.com/mixonic/ember-cli-deprecation-workflow
